graphql_compiler/compiler/ir_lowering_match/utils.py

```diff
@@ -237,50 +242,54 @@ def insert(self, optional_root_locations_path, complex_optional_roots):
-        if optional_root_location not in complex_optional_roots:
-            # Simple optionals are ignored.
-            # There should be no copmplex optionals after a simple optional.
-            if len(optional_root_locations_path) != 1:
-                raise AssertionError(u'Encountered simple optional root location {} at first '
-                                     u'position along path, but further locations are present. '
-                                     u'This should not happen: {}'
+        encountered_simple_optional = False
+        parent_location = self._root_location
+        for optional_root_location in optional_root_locations_path:
+            if encountered_simple_optional:
+                raise AssertionError(u'Encountered simple optional root location {} in path, but'
+                                     u'further locations are present. This should not happen: {}'
                                      .format(optional_root_location, optional_root_locations_path))
-            return
 
-        if optional_root_location not in self:
-            # Current Location does not exist as a child.
-            # Insert it and point it to an empty trie.
-            self[optional_root_location] = OptionalTraversalTrie()
+            if optional_root_location not in self._location_to_children:
+                # Simple optionals are ignored.
+                # There should be no complex optionals after a simple optional.
+                encountered_simple_optional = True
 
-        # Recurse on the suffix of the path.
-        child_trie_node = self[optional_root_location]
-        child_trie_node.insert(optional_root_locations_path[1:], complex_optional_roots)
+            else:
+                self._location_to_children[parent_location].add(optional_root_location)
+                parent_location = optional_root_location
 
-    def get_all_rooted_subtrees_as_lists(self):
+    def get_all_rooted_subtrees_as_lists(self, start_location=None):
         """Return a list of all rooted subtrees (each as a list of Location objects)."""
-        if self == {}:
-            # Empty trie only retuens a singleton list containing the null set (as an empty list).
+        if start_location is not None and start_location not in self._location_to_children.keys():
+            raise AssertionError(u'received invalid start_location {} that was not present '
+                                 u'in complex_optional_roots: {}'
+                                 .format(start_location, self._location_to_children.keys()))
+
+        if start_location is None:
+            start_location = self._root_location
+
+        if len(self._location_to_children[start_location]) == 0:
+            # Node with no children only returns a singleton list containing the null set.
             return [[]]
 
+        current_children = sorted(self._location_to_children[start_location], key=lambda l: repr(l))
```

Comment:
please don't sort by `repr` output -- this is expensive, non-obvious and very likely unstable across compiler versions